### DISCUSSION TITLE
set printer may has a bug

### DISCUSSION THREAD

**Discussion by qiushuitian** [infinum/floggy#37] [General]
/// Set custom printer on logger.
      set printer(LogPrinter? printer) {
          if (printer == null) {
            return;
          }

        clearListeners();

        _printer = printer;
        parent!._logStream.listen(_printer!.onLog); 
    }

does set printer will set the parent's printer?  ist it be this:

    _logStream.listen(_printer!.onLog);

**Comment by lukaknezic** [comment]
Hey @qiushuitian.

This is wanted behaviour, since you can see that `parent` is actually getter, that will return parent of loggy, or if that loggy doesn't have a parent it will return itself so it will actually act as if you called `this._logStream.listen`.

**Reply by qiushuitian** [reply]
Sometimes it effect too much. 

      final logger1 = Loggy("someName1");
      logger1.info("1111111");
      final logger2 = Loggy("someName2");
      logger1.info("2222222");
      logger1.printer = Printer1();
      logger1.info("3333333");
      logger2.printer = Printer2();
      logger1.info("4444444");
      logger2.info("5555555");
      print("done");

it will print:

    I/flutter (32680): RootPrinter: 1111111
    I/flutter (32680): RootPrinter: 2222222
    I/flutter (32680): Printer1: 3333333
    I/flutter (32680): Printer2: 4444444
    I/flutter (32680): Printer2: 5555555
    I/flutter (32680): done


it means logger2 set printer affect logger1.  Awalys I wan't a logger has itself printer.

**Reply by lukaknezic** [reply]
@qiushuitian Ooh okay, I think in that case you are looking for `Loggy.detached`. 

That loggy is detached from it's parent and can have it's own printer and rules (It won't be filtered or affected by setup in `Loggy.init`). 
So just changing your example to: 
```dart
final logger1 = Loggy.detached("someName1");
logger1.info("1111111");
final logger2 = Loggy.detached("someName2");
```
Will give this output:
```
Printer1: [I] someName1: 3333333
Printer1: [I] someName1: 4444444
Printer2: [I] someName2: 5555555
done
```
_There is no RootPrinter so first 2 calls are not shown_

You can use helper function `detachedLoggy(String name, {LoggyPrinter? logPrinter})` where you can specify printer for loggy as well, since `Loggy.detached(String name)` doesn't have printer attached to it by default, and it cannot use parent's printer.
